petric.py

```diff
@@ -88,26 +87,58 @@ def __call__(self, algo: Algorithm):
         log.debug("...logged")
 
 
+class QualityMetrics(ImageQualityCallback):
+    """From https://github.com/SyneRBI/PETRIC/wiki#metrics-and-thresholds"""
+    def __init__(self, reference_image, backround_mask, foreground_mask, **kwargs):
+        super().__init__(reference_image, **kwargs)
+        self.background = np.where(backround_mask == 1)
+        self.foreground = np.where(foreground_mask == 1)
+
+    def __call__(self, algorithm):
+        from skimage import metrics as sm
+
+        iteration = algorithm.iteration
+        if iteration % algorithm.update_objective_interval != 0 and iteration != algorithm.max_iteration:
+            return
+        test_image = algorithm.x # CIL or SIRF ImageData
+
+        # # (0) objective value
+        # objective = algorithm.get_last_objective(return_all=False)
+        # self.tb_summary_writer.add_scalar('objective', objective, iteration)
+
+        test_im_arr, ref_im_arr = test_image.as_array(), self.reference_image.as_array()
+
+        for filter_name, filter_func in self.filter.items():
+            if filter_func is not None:
+                test_im, ref_im = map(filter_func, (test_im_arr, ref_im_arr))
+
+            # (1) global metrics & statistics
+            norm = ref_im[self.background].mean()
+            self.tb_summary_writer.add_scalar(
+                f"RMSE_foreground{filter_name}",
+                np.sqrt(sm.mean_squared_error(test_im[self.foreground], ref_im[self.foreground])) / norm, iteration)
+            self.tb_summary_writer.add_scalar(
+                f"RMSE_background{filter_name}",
+                np.sqrt(sm.mean_squared_error(test_im[self.background], ref_im[self.background])) / norm, iteration)
+
+            # (2) local metrics & statistics
+            for roi_name, roi_inds in self.roi_indices.items():
+                # AEM not to be confused with MAE
+                self.tb_summary_writer.add_scalar(f"AEM_VOI_{roi_name}{filter_name}",
+                                                  np.abs(test_im[roi_inds].mean() - ref_im[roi_inds].mean()) / norm,
+                                                  iteration)
```

Comment:
@KrisThielemans I could not understand https://github.com/SyneRBI/PETRIC/wiki#metrics-and-thresholds, atm I assume it means:

1. $RMSE_{foreground} = RMSE(c_W, r_W) / MEAN(r_B)$
   - surely using $MEAN(r)$ is better than $MEAN(r_B)$?
2. $RMSE_{background} = RMSE(c_B, r_B) / MEAN(r_B)$
  - why do we care about this?
3. $AEM_{VOI i} = |MEAN(c_{Ri}) - MEAN(r_{Ri})| / MEAN(r_B)$
  - not to be confused with $MAE_{VOI i}$

Reply:
This is correct